avocado/core/__init__.py

```diff
@@ -11,3 +11,26 @@
 #
 # Copyright: Red Hat Inc. 2013-2014
 # Author: Lucas Meneghel Rodrigues <lmr@redhat.com>
+
+
+from .future.settings import settings as future_settings
+from .output import BUILTIN_STREAMS, BUILTIN_STREAM_SETS
+
+
+def register_core_options():
+    streams = (['"%s": %s' % _ for _ in BUILTIN_STREAMS.items()] +
+               ['"%s": %s' % _ for _ in BUILTIN_STREAM_SETS.items()])
+    streams = "; ".join(streams)
+    help_msg = ("List of comma separated builtin logs, or logging streams "
+                "optionally followed by LEVEL (DEBUG,INFO,...). Builtin "
+                "streams are: %s. By default: 'app'" % streams)
+    future_settings.register_option(section='core',
+                                    key='show',
+                                    key_type=lambda x: x.split(','),
+                                    metavar="STREAM[:LVL]",
+                                    nargs='?',
+                                    default=['app'],
+                                    help_msg=help_msg)
+
+
+register_core_options()
```

Comment:
Usually, I try to avoid adding code in `__init__.py`. 

From the "The Hitchhiker’s Guide to Python!":

> A commonly seen issue is to add too much code to `__init__.py` files. When the project complexity grows, there may be sub-packages and sub-sub-packages in a deep directory structure. In this case, importing a single item from a sub-sub-package will require executing all `__init__.py` files met while traversing the tree.
> 
> Leaving an `__init__.py` file empty is considered normal and even a good practice, if the package’s modules and sub-packages do not need to share any code.

Unless I'm missing something, I'm not sure if this code needs to be shared with all modules. It looks like a code that needs to be executed only once. So is this the case, would suggest adding this to `app.py` or `core.py`

Reply:
@beraldoleal  the problem with adding to another file, such as `app.py` or `core.py`, is that it'd require users (say, of the job API), to manually import those modules. Do you have a suggestion that doesn't incur on that usability penalty?

Reply:
@clebergnu, probably I'm missing something.

I thought that the idea was to make this "registration" transparent to the user, right? So we could call this method at the `__init__` or `__enter__` methods of the `Job()` class, right?

Reply:
For everything that is related to a job, what you suggest is the best course of action. That's why we have `avocado/core/job.py:register_job_options()`. The problem with this option, is that it's *not* related to a job, it's a foundation to many different use cases and commands.

One other approach is to call this explicitly in every single place it may be needed.  For instance, it may be needed by `avocado variants`, and it'd have to be called there.  I don't fancy that, though.

Reply:
Yes, I think that understand the differences between `register_job_options()` and `register_core_options()`, and my comment took that into consideration.

What sounds like a natural course of action to me would be defining the method `register_job_options()` in someplace like `core.py` (for instance) and call it explicitly in every place it may be needed, as long as the user does not need to call. For me, it is better when reading coding.

In that case, this could be called from `__init__()` and/or  `__enter__()` from `Job()`, from`variants.py`, and any other place (as long as it is called only once).

Even if this is not part of the job, this is necessary to run the job, and in this case, IMO, explicitly is better than implicitly:

```python
class Job:
  def __init__(self):
    register_core_options()
    register_job_options()
```

Please, keep in mind that this is just a suggestion and what I would try.

Reply:
> Yes, I think that understand the differences between `register_job_options()` and `register_core_options()`, and my comment took that into consideration.
> 
> What sounds like a natural course of action to me would be defining the method `register_job_options()` in someplace like `core.py` (for instance) and call it explicitly in every place it may be needed, as long as the user does not need to call. For me, it is better when reading coding.
> 

So those methods would need to be able to be called multiple times, but only perform the real action (registering options) once.  This would require a global state holder, or a change to `settings.register_option` to not care about trying to register multiple times.

> In that case, this could be called from `__init__()` and/or `__enter__()` from `Job()`, from`variants.py`, and any other place (as long as it is called only once).
> 

Like I said before, this would need extra logic, to make sure "it is called only once". IMO it's *more* error prone than having one single place where it's called and executed.

> Even if this is not part of the job, this is necessary to run the job, and in this case, IMO, explicitly is better than implicitly:
> 
> ```python
> class Job:
>   def __init__(self):
>     register_core_options()
>     register_job_options()
> ```
> 

Doing it inside the Job class is not possible as it is.  You can easily see how it'd conflict when a second Job is instantiated. And AFAICT there should be no reason not to allow Job API users to instantiate more than one Job. That's why `register_job_options()` currently done at `avocado.core.job`.  The same logic applies to core, only that core is supposed to have an effect that would be available to a much wider audience.

> Please, keep in mind that this is just a suggestion and what I would try.

Sure.  Let me know if you think the extra work of keeping the global state is worth it.  I clearly don't think it is.  And if you do, maybe prototyping a solution would be a good course of action, so that we can change the path of this PR based on more evidence.

Thanks!

Reply:
@clebergnu I see your points.

As I said, I would try to avoid adding things to `__init__` because I'm naturally used to avoid it.

But I understand now that this would require a little extra effort to avoid double registration.

So, let's move on and if we realize that __init__ is getting bigger, we can move from there in the future.

Reply:
@beraldoleal let me know if there's anything else we need to address on this PR. Thanks.